File: Problems/Codeforces/Division 3/600's/661/Weights Division(hard).py
from collections import deque
import sys
input = sys.stdin.readline
def print(val):
    sys.stdout.write(str(val) + '\n')
    
def prog():
    for _ in range(int(input())):
        n,s = map(int,input().split())
        adj = [[] for i in range(n)]
        for i in range(n-1):
            u,v,w,c = map(int,input().split())
            u -= 1
            v -= 1
            adj[u].append([v,w,c])
            adj[v].append([u,w,c])
        leaves_in_subtree = [0]*n
        stack = deque([0])
        visited = [0]*n
        parent = [0]*n
        while stack:
            curr = stack[-1]
            if not visited[curr]:
                visited[curr] = 1
                for neighbor in adj[curr]:
                    if neighbor[0] != parent[curr]:
                        parent[neighbor[0]] = curr
                        stack.append(neighbor[0])
            else:
                if len(adj[curr]) == 1:
                    leaves_in_subtree[curr] = 1
                else:
                    for neighbor in adj[curr]:
                        leaves_in_subtree[curr] += leaves_in_subtree[neighbor[0]]
                stack.pop()
 
        weight_delta1 = []
        weight_delta2 = []
        total_weight = 0
        for u in range(1, n):
            for neighbor in adj[u]:
                if neighbor[0] == parent[u]:
                    if neighbor[2] == 1:
                        weight = neighbor[1]
                        total_weight += weight*leaves_in_subtree[u]
                        while weight:
                            weight_delta1.append((weight - weight//2)*leaves_in_subtree[u])
                            weight //= 2
                    else:
                        weight = neighbor[1]
                        total_weight += weight*leaves_in_subtree[u]
                        while weight:
                            weight_delta2.append((weight - weight//2)*leaves_in_subtree[u])
                            weight //= 2
        weight_delta1.sort(reverse = True)
        weight_delta2.sort(reverse = True)
 
        cost = 0
        idx1 = 0
        idx2 = 0
        while total_weight > s:
            if idx1 != len(weight_delta1) and total_weight - weight_delta1[idx1] <= s:
                cost += 1
                break;
            elif idx1 < len(weight_delta1) -1 and (idx2 == len(weight_delta2) or\
            weight_delta2[idx2] < weight_delta1[idx1] + weight_delta1[idx1 + 1]):
                total_weight -= weight_delta1[idx1] + weight_delta1[idx1 + 1]
                idx1 += 2
                cost += 2
            else:
                total_weight -= weight_delta2[idx2]
                idx2 += 1
                cost += 2
        print(cost)
prog()

File: Problems/Codeforces/Division 3/600's/661/661/Weights_Division(Hard).cpp
#include<bits/stdc++.h>
using namespace std;

struct edge {
	int v, w, c;
};

int main() {
	ios_base::sync_with_stdio(0); cin.tie(0);
	int t; cin >> t;
	while (t--) {
		long long n, s; cin >> n >> s;
		vector<vector<edge>> adj(n);
		for (int i = 0; i < n - 1; i++) {
			int u, v, w, c; cin >> u >> v >> w >> c;
			u--;
			v--;
			adj[u].push_back({ v,w,c });
			adj[v].push_back({ u,w,c });
		}

		vector<int> leaves_in_subtree(n);
		stack<int> sk; sk.push(0);
		vector<bool> visited(n);
		vector<int> parent(n);
		while (!sk.empty()) {
			int curr = sk.top();
			if (!visited[curr]) {
				visited[curr] = 1;
				for (auto neighbor : adj[curr]) {
					if (neighbor.v != parent[curr]) {
						parent[neighbor.v] = curr;
						sk.push(neighbor.v);
					}
				}
			}
			else {
				if (adj[curr].size() == 1) {
					leaves_in_subtree[curr] = 1;
				}
				else {
					for (auto neighbor : adj[curr]) {
						leaves_in_subtree[curr] += leaves_in_subtree[neighbor.v];
					}
				}
				sk.pop();
			}
		}

		vector<long long> weight_delta1;
		vector<long long> weight_delta2;
		long long total_weight = 0;
		for (int u = 1; u < n; u++) {
			for (auto neighbor : adj[u]) {
				if (neighbor.v == parent[u]) {
					if (neighbor.c == 1) {
						long long weight = neighbor.w;
						total_weight += weight * leaves_in_subtree[u];
						while (weight) {
							weight_delta1.push_back((weight - weight / 2) * leaves_in_subtree[u]);
							weight /= 2;
						}
					}
					else {
						long long weight = neighbor.w;
						total_weight += weight * leaves_in_subtree[u];
						while (weight) {
							weight_delta2.push_back((weight - weight / 2) * leaves_in_subtree[u]);
							weight /= 2;
						}
					}
				}
			}
		}

		sort(weight_delta1.begin(), weight_delta1.end(), greater<long long>());
		sort(weight_delta2.begin(), weight_delta2.end(), greater<long long>());

		vector<long long> psum1(weight_delta1.size() + 1);
		vector<long long> psum2(weight_delta2.size() + 1);
		for (int i = 0; i < weight_delta1.size(); i++) {
			psum1[i + 1] = psum1[i] + weight_delta1[i];
		}
		for (int i = 0; i < weight_delta2.size(); i++) {
			psum2[i + 1] = psum2[i] + weight_delta2[i];
		}

		int min_cost = 1e9;
		for (int i = 0; i < psum2.size(); i++) {
			int cost = 2 * i;
			long long needed = total_weight - psum2[i] - s;
			if (needed <= 0) {
				min_cost = min(min_cost, cost);
				break;
			}
			else {
				int L = 0;
				int R = psum1.size() - 1;
				while (L != R) {
					int m = (L + R) / 2;
					if (psum1[m] >= needed) {
						R = m;
					}
					else {
						L = m + 1;
					}
				}
				if (needed <= psum1[L]) {
					cost += L;
					min_cost = min(min_cost, cost);
				}
			}
		}

		cout << min_cost << '\n';

	}
}